docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,519 @@
+# ADR 008: RPC Client Event Subscription Mechanism
+
+## Changelog
+
+* 2020-07-23: Initial draft
+
+## Context
+
+The [Tendermint Light Client](../../light-client/) is one of many important
+applications that will make use of the [RPC client](../../rpc/) to query
+Tendermint full nodes for information relating to the blockchain.
+
+Tendermint servers (e.g. full nodes) provide an [event
+subscription][tm-event-subs] RPC endpoint to allow clients to receive
+notifications of specific events as they happen (currently via a WebSockets
+connection). We need to expose this subscription mechanism through the RPC client.
+
+In order to achieve this, we need:
+
+1. An ergonomic interface for the RPC client that allows developers to subscribe
+   to events produced by specific queries. Specifically, this interface must:
+   1. Offer **subscription** functionality, where:
+      1. A single **subscription** takes place to the **events** produced by a
+         **query** (the [PEG] for which, at the time of this writing, is located
+         [here][query-peg]).
+      2. Callers must be able to create multiple distinct subscriptions.
+      3. A subscription must offer an interface to allow for iteration over
+         incoming events only relevant to that subscription (i.e. it should
+         produce an **event stream**).
+      4. It must be possible to, from outside of the RPC client, receive events
+         from multiple subscriptions' event streams concurrently without
+         interfering with/blocking each other.
+      5. It must be possible to, from outside of the RPC client, handle
+         subscriptions' event streams without blocking other RPC operations.
+   2. Offer the ability to **unsubscribe** from specific queries (i.e. to
+      terminate specific subscriptions).
+2. An appropriate concurrency model for drivers of the transport layer that
+   allows the transport layer to operate independently of consumers of
+   subscriptions. This is so that consumers don't block transport layer
+   activities and vice-versa.
+
```

Comment:
Would be helpful if we also had context on what currently exists in the code base and what the issues are.